Fix for #879

zemian opened the pull request.

vlsi commented:
I wonder if #898  can help testing this change

zemian commented:
Hi Vlsi, yeah testing this is little difficult since I dont' see the PG conn properties being saved anywhere after the conn object is made. I did however did a manual break point check to verify the timeout value been set in my IDE and it seems to work.

I am not familiar with "netcrusher" mentioned in other issue. Also the link mentioned there gives a 404 page!

I do see one more improvement can be made is that since PGProperty.LOGIN_TIMEOUT always default with "0", then it means we don't need NULL check in Driver.java line 645. If I remove this, your coverage might get back to previous stat. Let me know if you want this or not.

davecramer commented:
@zemian the code is fine the way it is, thanks for the PR. I hope nobody really uses a float for a timeout!

jorsol commented:
I'm curious, what happen when a negative number in introduced?
Can we have a check to allow only positive?

vlsi commented:
>the code is fine the way it is, thanks for the PR. I hope nobody really uses a float for a timeout!

I wonder what is the declared/documented semantics for loginTimeout=0.

zemian commented:
@jorsol Current Driver.java code check timeout value with "<= 0" cond to make direct conn. So negative value will be same as 0.

@davecramer Great! Glad I can help!

davecramer commented:
@vlsi I think we can do whatever we want with it which IMO would be to disable it when it is set to 0

vlsi commented on pgjdbc/src/main/java/org/postgresql/Driver.java:
```diff
@@ -640,15 +640,20 @@ private static String database(Properties props) {
       }
     }
-    return (long) DriverManager.getLoginTimeout() * 1000;
+    if (result == 0) {
```
Pease use `org.postgresql.PGProperty#isPresent` to check property presence.
`result != 0` is not a proper check as user might specify `loginTimeout=0` explicitly to disable login timeout.

zemian replied:
Ah was aware that method. I can update, but there might be slight behavior difference though. As you said this allow explicit 0 set by property to not check DriverManager. Basically  any property value is given, even 0, DriverManager value will not be used. If you can confirm this, I will make the change as well.

vlsi commented on pgjdbc/src/main/java/org/postgresql/Driver.java:
```diff
@@ -640,15 +640,20 @@ private static String database(Properties props) {
     if (timeout != null) {
       try {
-        return (long) (Float.parseFloat(timeout) * 1000);
+        result = Integer.parseInt(timeout);
```
Are there good reasons to convert from `float` to `int` parsing?
I agree it is a bit esoteric, however I would refrain from altering things without clear justification.

zemian replied:
Hi vlsi, I explained the reason in original issue #879 comments. If you still think we shouldn't change, then I will revert it back.

vlsi replied:
As far as I understand, the only reason is "I assume loginTimeout value should always be integer in secs".

There might be even a good reason to use float values: ability to specify sub-second timeouts while still being in line with DriverManager's `second` units.

Having said that, I thing we should refrain from altering the code.

zemian commented:
@vlsi  The latest commit should addressed your concerns. Let me know what you think.

vlsi commented on pgjdbc/src/main/java/org/postgresql/Driver.java:
```diff
@@ -640,15 +640,18 @@ private static String database(Properties props) {
         LOGGER.log(Level.WARNING, "Couldn't parse loginTimeout value: {0}", timeout);
       }
+    } else {
+      result = (long) DriverManager.getLoginTimeout() * 1000;
```
Do you mean to use the `DriverManager.getLoginTimeout` only in case of option not being present in the props?
I think it makes sense to revert to default case in case of "parse failure" as well.

davecramer replied:
Agreed

zemian replied:
> Do you mean to use the DriverManager.getLoginTimeout only in case of option not being present in the props?
Yes.

> I think it makes sense to revert to default case in case of "parse failure" as well.
For default case, do you mean in the catch block you prefer to use `PGProperty.LOGIN_TIMEOUT.getDefaultValue()`?

davecramer replied:
Actually thinking about this. If parsing fails we should probably throw an exception as the user probably wanted to set the timeout but somehow failed to set it properly. By not failing we are silently ignoring an error. I don't think that is what users would want.

zemian replied:
Hi Dave, sounds like a good idea.

I do have one suggestion though. If we go with throwing exception, then we are basically doing very similar to existing PGProperty.getInt() method, except it will be a float in return instead. Do you want to add a new getFloat method in PGProperty class for reuse in future? Or you like to just keep the changes in Driver.java for now?

zemian commented:
Hm... I am not sure I understand the reason for failed checks here on the last commit, which only has a javadoc misspelled word! If there is anything I can help let me know.

zemian commented:
@vlsi, after some testing, I noticed one odd behavior with `PGProperty.isPrensent()` checks. Due to the way `Driver.connect(url, info)` is using the "info" input parameter as `java.util.Properties#defaults` only, the `PGProperty.isPrensent() ` will only return true if "loginTime" is set from URL, and not when directly pass in as the "info" Properties object!

What I mean is the following test case work as expected with override (logintTime will be 3):
```
        DriverManager.setLoginTimeout(5);
        String url = "jdbc:postgresql://localhost/test?loginTimeout=3";
        try(Connection conn = DriverManager.getConnection(url)) {
            System.out.println(conn);
        }
```

But this case will NOT override (loginTime will be 5 since PGProperty.isPrensent() return false in this case)!
```
        DriverManager.setLoginTimeout(5);
        String url = "jdbc:postgresql://localhost/test";
        Properties info = new Properties();
        info.setProperty("loginTimeout", "3");
        try(Connection conn = DriverManager.getConnection(url, info)) {
            System.out.println(conn);
        }
```

I think this can be very confusing for users!

zemian commented:
I now have the PGProperty.isPresent check problem fixed. I also found an existing LoginTimeoutTest that I added couple use cases there. Please review and let me know if these are good with you.

vlsi commented on pgjdbc/src/main/java/org/postgresql/Driver.java:
```diff
@@ -523,7 +526,10 @@ public boolean jdbcCompliant() {
-    Properties urlProps = new Properties(defaults);
+    Properties urlProps = new Properties();
+    if (defaults != null) {
+      urlProps.putAll(defaults);
```
I've created "Properties, defaults, PR 900, etc" discussion on the pgjdbc-jdbc mailing list.
I would appreciate @whitingjr and/or @alexismeneses to step in as well.

Just in case: `urlProps.putAll(defaults);` will use `defaults.entrySet()` and it will ignore "parent properties" of a default object.

It happens so that this particular `Properties defaults` is initialized with "parent Properties" object (see https://github.com/pgjdbc/pgjdbc/blob/ede2da1898aee4e503e1041c133dbf95322fa4a8/pgjdbc/src/main/java/org/postgresql/Driver.java#L220 ), so I expect this change will break `getDefaultProperties();`.

vlsi commented on pgjdbc/src/main/java/org/postgresql/Driver.java:
```diff
@@ -639,16 +645,14 @@ private static String database(Properties props) {
+    if (PGProperty.LOGIN_TIMEOUT.isPresent(props)) {
+      result = (long) (PGProperty.LOGIN_TIMEOUT.getFloat(props) * 1000);
+    } else {
+      result = (long) (DriverManager.getLoginTimeout() * 1000);
```
I think the proper code would be to express `1000` as a long as in `1000L` otherwise there's a hidden silent integer overflow: https://stackoverflow.com/a/7215432/1261287

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/LoginTimeoutTest.java:
```diff
@@ -81,8 +82,39 @@ public void testBadTimeout() throws Exception {
+    DriverManager.setLoginTimeout(5);
+    Connection conn = DriverManager.getConnection(TestUtil.getURL(), props);
+    conn.close();
+    DriverManager.setLoginTimeout(0);
```
Does this test actually test anything?
Would you please add some assertions?

davecramer replied:
@zemian can you respond to this ?

zemian replied:
Hi Dave, as I was saying, I did replied on the Conversation tab regarding this (https://github.com/pgjdbc/pgjdbc/pull/900#issuecomment-341902369)... maybe it's not directly linked here. So let me re-paste it here for your reference:

> Hi @vlsi , I have put in a change that addressed your concerns now. Please review
>
>As far as test assertions. I have done similar to your existing LoginTimeoutTest, which contains few tests already doing similar thing (no assertion there), it merely ensure connection object are obtained correctly with timeout settings set. I doing same but with DriverManager.setLoginTimeout() instead. However I just now added one more new test for "testTimeoutOccursUsingDriverManagerLoginTimout", which also test just like existing one but with the DriverManager.setLoginTimeout(). This test has the assertion for real timeout. I think it's good validation for the change in this PR.

gongmingbin replied:
or just remove this test?

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/LoginTimeoutTest.java:
```diff
@@ -81,8 +82,39 @@ public void testBadTimeout() throws Exception {
+    props.setProperty("loginTimeout", "0");
     Connection conn = DriverManager.getConnection(TestUtil.getURL(), props);
     conn.close();
+    DriverManager.setLoginTimeout(0);
```
Same as the above. Please add assertions.

davecramer replied:
and this

zemian replied:
Same respond as above.

zemian commented:
1) Hi @vli, sorry for the long delay update to this PR. I read your comments and to the post on mailing list as well. I agree that loading Properties default should be consistent and should not have two ways of loading it. 

I see @davecramer stated he prefers "new Properties(defaultProps)" style of loading Properties. (I personally prefer the other way because the "new Properties(defaultProps)" style is harder to debug.) So I can revert this change in PR.

However if I do revert above change, that means I can't use "if (PGProperty.LOGIN_TIMEOUT.isPresent(props))" check. Can you suggest how we should go about detecting this condition?

2) Yes I agree that I should use `1000L` where needed. I will update this.

3) As for improving the test. The problem I faced was that testing "loginTimeout" value is tricky since there is not good way to simulate this in test env where the system will give timeout. And I don't see a way to get "loginTimeout" value for assertion before connection is made. So the only thing I can perform at this moment is ensure using the "loginTimeout" parameter will not break in getting the connection object. If you have suggestion please let me know.

zemian commented:
Hi @vlsi , I have put in a change that addressed your concerns now. Please review

As far as test assertions. I have done similar to your existing LoginTimeoutTest, which contains few tests already doing similar thing (no assertion there), it merely ensure connection object are obtained correctly with timeout settings set. I doing same but with DriverManager.setLoginTimeout() instead. However I just now added one more new test for "testTimeoutOccursUsingDriverManagerLoginTimout", which also test just like existing one but with the DriverManager.setLoginTimeout(). This test has the assertion for real timeout. I think it's good validation for the change in this PR.

davecramer commented:
LGTM
